cism runoff will be now routed to ocn via mosart

mvertens opened the pull request:
This PR enables CISM runoff to be routed to the ocean via mosart.
- All runoff from CISM will be routed directly to the outlet points
- New fields will be advertised in the mosart cap
      call fldlist_add(fldsFrRof_num, fldsFrRof, 'Forr_rofl_glc')
      call fldlist_add(fldsFrRof_num, fldsFrRof, 'Forr_rofi_glc')

Issues resolved:
- Resolves: #102
- Resolves: #92
- Resolves: #101

TODOs:
- [x] This PR must come in after #76 since it is built on top of those changes.
- [x] Merge in mosart1.1.01 into this branch (shouldn't change much, since this was built on top of #76)
- [x] This PR must also be accompanied by CISM, CMEPS, ccs_config PRs.
   - https://github.com/ESCOMP/CMEPS/pull/463
   - https://github.com/ESCOMP/CISM-wrapper/pull/100
   - https://github.com/ESMCI/ccs_config_cesm/pull/174
- [x] Ran mosart test suite on derecho and compared with mosart1_0_49 baseline
       Results are bfb (other than new fields on history files) and new baselines created mosart1.1.2
- [x] Some validation has been carried out. 
       See https://github.com/ESCOMP/CISM-wrapper/issues/64. Determine if more validation is needed.
       @billsacks @katec - are we okay with the current state of the validation?

mvertens commented on src/riverroute/mosart_driver.F90:
```diff
@@ -0,0 +1,927 @@
+         close(unitn)
+      end if
+
+      call mpi_bcast (finidat, len(finidat), MPI_CHARACTER, 0, mpicom_rof, ier)
```
Originally I thought this was a good idea. Looking at this some more I found that ESMF_VMBroadcast only accepts arrays. That means that an array of length 1 needs to be created and then a copy of finidat needs to be made to that and then broadcast and then copied back. I think this will make the code more complicated and harder to read. 
At this point I recommend staying with native mpi calls. I'm happy to talk about this some more.

mvertens replied:
See issue https://github.com/ESCOMP/MOSART/issues/100.

mvertens commented on src/riverroute/mosart_budget_type.F90:
```diff
@@ -0,0 +1,261 @@
+      end do
+
+      tmp_in = tmp_in*1e-6_r8 !convert to million m3
+      call shr_mpi_sum(tmp_in, tmp_glob, mpicom_rof, 'mosart global budget', all=.false.)
```
However, ESMF_VMAllReduce only accepts one-dimensional arrays. tmp_in and tmp_glob are 2d-arrays. So copies will need to be made and 6 calls will need to be done and then copies need to be made. Frankly, I am worried that moving to only ESMF MPI communication will make the code harder to understand since it really supports only a very small subset of MPI communications. And I think it is not a good idea to mix ESMF calls in the code with MPI calls. Happy to talk about this some more.

mvertens replied:
See issue https://github.com/ESCOMP/MOSART/issues/100.

mvertens commented:
> I recommend removing all direct references to mpi and using ESMF_VM calls instead - this future proofs the code a little in case the mpi library is replaced with some other comm library on some future hardware - if all our calls are through the esmf layer then only esmf needs to be updated.

I have concerns about this (also listed below). 
-  ESMF_VMAllReduce only accepts one-dimensional arrays. tmp_in and tmp_glob are 2d-arrays. So copies will need to be made and 6 calls will need to be done and then copies need to be made. Frankly, I am worried that moving to only ESMF MPI communication will make the code harder to understand since it really supports only a very small subset of MPI communications. And I think it is not a good idea to mix ESMF calls in the code with MPI calls. Happy to talk about this some more.
- Using ESMF_VMBroadcast - originally I thought this was a good idea. Looking at this some more I found that ESMF_VMBroadcast only accepts arrays. That means that an array of length 1 needs to be created and then a copy of finidat needs to be made to that and then broadcast and then copied back. I think this will make the code more complicated and harder to read.
At this point I recommend staying with native mpi calls. I'm happy to talk about this some more.

mvertens commented on src/riverroute/mosart_budget_type.F90:
```diff
@@ -0,0 +1,261 @@
+   use mosart_vars,        only: re, spval, barrier_timers, iulog, mainproc, npes, iam, mpicom_rof
+   use mosart_data,        only: ctl, Tctl, Tunit, TRunoff, Tpara
+   use mosart_timemanager, only: get_nstep, get_curr_date
+   use mpi
```
See issue https://github.com/ESCOMP/MOSART/issues/100

mvertens commented on src/riverroute/mosart_control_type.F90:
```diff
@@ -0,0 +1,1238 @@
+       write(iulog,*) subname,'ERROR this mask lt 1 ',minval(this%mask),maxval(this%mask)
+       call shr_sys_abort(subname//' ERROR this mask')
+    endif
+    call shr_mpi_sum(larea, this%totarea, mpicom_rof, 'mosart totarea', all=.true.)
```
See issue https://github.com/ESCOMP/MOSART/issues/100.

mvertens commented on src/riverroute/mosart_control_type.F90:
```diff
@@ -0,0 +1,1238 @@
+    if (mainproc) then
+       write(iulog,*) 'total runoff cells numr  = ',numr
+    endif
+    call mpi_barrier(mpicom_rof,ier)
```
See issue https://github.com/ESCOMP/MOSART/issues/100.

mvertens commented on src/riverroute/mosart_control_type.F90:
```diff
@@ -0,0 +1,1238 @@
+          write(iulog,'(2a,i9,a,i9,a,i9,a,i9)')' mosart decomp info',&
+               ' proc = ',iam,' begr = ',begr,' endr = ',endr,' numr = ',lnumr
+       endif
+       call mpi_barrier(mpicom_rof,ier)
```
See issue https://github.com/ESCOMP/MOSART/issues/100.

mvertens commented on src/riverroute/mosart_driver.F90:
```diff
@@ -0,0 +1,927 @@
+   use mosart_io          , only : ncd_pio_openfile, ncd_inqdid, ncd_inqdlen, ncd_pio_closefile, ncd_decomp_init, &
+                                   pio_subsystem
+   use pio                , only : file_desc_t
+   use mpi
```
Done.

slevis-lmwg reviewed (approved):
@mvertens I saw your post, so I will go ahead and approve. I will coordinate the merge with @ekluzek to make sure that I handle this comment correctly:
"This PR must also be accompanied by CISM, CMEPS, ccs_config PRs (to be listed here as they are generated)."

mvertens commented:
@slevis-lmwg - I still need to run the mosart test suite and make sure that everything is bfb. So no hurry to approve just yet.

slevis-lmwg commented:
> @slevis-lmwg - I still need to run the mosart test suite and make sure that everything is bfb. So no hurry to approve just yet.

@mvertens will you merge mosart1.1.01 to this PR or would you like me to do it and, if the latter, then before or after you run the test suite?

mvertens commented:
@slevis-lmwg - I'll merge mosart1.1.0.1 to this PR before I run the test suite.

mvertens commented:
@slevis-lmwg - I think this is ready for merging based on all the tests I've done. @jedwards4b - do you want to review again?

mvertens commented:
@billsacks @Katetc - are you both okay with the current level of validation?

jedwards4b reviewed (approved):
lgtm

billsacks commented:
Yes, I'm satisfied with the current level of validation. Thank you very much for your work on this @mvertens !!!

slevis-lmwg commented:
I have checked out this branch and submitted the test-suites on izumi and derecho:
`./run_sys_tests -s mosart -c mosart1.1.01_ctsm5.2.007 -g mosart1.1.02_ctsm5.2.007`

slevis-lmwg commented:
Oh, right, and i have confirmed that we need to complete https://github.com/ESCOMP/CTSM/issues/2590 for the tests to pass.

jedwards4b commented:
I think that there is going to be a bit of a chicken and egg problem here - you will need to use some PR branches to test some other PR's.

ekluzek commented on src/riverroute/mosart_physics.F90:
```diff
@@ -270,12 +279,6 @@ subroutine mainchannelRouting(nr, nt, theDeltaT)
-      else if(Tctl%RoutingMethod == 4) then
-         call Routing_DW(nr, nt, theDeltaT)
       else
          call shr_sys_abort( "mosart: Please check the routing method! There are only 4 methods available." )
```
The comment in RunoffMod.F90 should also be updated.

mvertens replied:
Done. However, there is no longer a routine called RunoffMod.F90.

ekluzek commented on src/riverroute/mosart_budget_type.F90:
```diff
@@ -203,23 +222,22 @@ subroutine check_budget(this, begr, endr, ntracers, dt)
          error_budget = .false.
          abserr = 0.0_r8
          relerr = 0.0_r8
-         this%beg_vol_glob(nt) = tmp_glob(1, nt)
```
Love that you are fixing #95 here.

ekluzek reviewed (changes requested):
@mvertens thanks for all your work here. I have some updates that @slevis-lmwg can do, if you aren't able to do them. Or if you prefer to do them go ahead. Just let us know how to handle it.

One thing @slevis-lmwg realized when he started testing is that the separate_glc2ocn_fluxes namelist item assumes that ctl%rof_from_glc is true, and hence the code won't work with an older CISM version. In order to test this seperately it would be good to add some error checking that they are both true, and if not to abort. This also allows us to test with that flag to false and make sure it doesn't change answers, which is a good verification that the other refactoring is b4b.

So let us know how you want to handle the changes and we can continue forward.

mvertens commented on src/riverroute/mosart_driver.F90:
```diff
@@ -162,17 +165,18 @@ subroutine mosart_read_namelist()
+         write(iulog,'(a)'   ) '   qgwl runoff option      = '//trim(qgwl_runoff_option)
+         write(iulog,'(a)'   ) '   mosart tracers          = '//trim(mosart_tracers)
+         write(iulog,'(a)'   ) '   mosart euler calc       = '//trim(mosart_euler_calc)
+         write(iulog,'(a,l1)') '   separate_glc2ocn_fluxes = ',separate_glc2ocn_fluxes
```
I have removed separate_glc2ocn_fluxes as a namelist variable - and simply have mosart send separate fluxes. The mediator now knows how to handle this in a separate PR and this added needles complexity at this point.

mvertens commented on src/riverroute/mosart_control_type.F90:
```diff
@@ -115,13 +123,49 @@ module mosart_control_type
+!========================================================================
+
+  subroutine init_tracer_names(this, mosart_tracers)
+
```
Done.

mvertens commented on src/cpl/nuopc/rof_import_export.F90:
```diff
@@ -82,9 +83,14 @@ subroutine advertise_fields(gcomp, flds_scalar_name, rc)
     call fldlist_add(fldsFrRof_num, fldsFrRof, 'Forr_rofl')
     call fldlist_add(fldsFrRof_num, fldsFrRof, 'Forr_rofi')
+    if (separate_glc2ocn_fluxes) then
+      call fldlist_add(fldsFrRof_num, fldsFrRof, 'Forr_rofl_glc')
```
separate_glc2ocn_fluxes has now been removed.

mvertens replied:
See above.

mvertens commented on src/riverroute/mosart_control_type.F90:
```diff
@@ -115,13 +123,49 @@ module mosart_control_type
+
+    ! Local variables
+    integer :: nt
+    character(len=*),parameter :: subname = '(mosart_control_type: init_tracer_names)'
```
subname is used extensively in mosart and I think that the implementation of this change should occur in a follow up PR. This PR is critically needed for CESM3.0 as well as NorESM2.5  - so I don't see this as being necessary for this PR.
@billsacks - can you please comment on this as well.

jedwards4b replied:
I agree with @mvertens, this change can be deferred to another PR.

billsacks replied:
Agreed. Let's not hold up this PR with non-critical refactoring like this that can be addressed later.

ekluzek replied:
Sounds good.

mvertens commented on cime_config/namelist_definition_mosart.xml:
```diff
@@ -288,10 +288,24 @@
+    <desc>
+      Default: .true.
+      If .true., glc2ocn fluxes that are passed through mosart will be sent
+      as a separate fields to the mediator.
```
This namelist has been removed - so this is no longer needed.

mvertens commented on src/cpl/nuopc/rof_import_export.F90:
```diff
@@ -222,6 +230,15 @@ subroutine realize_fields(gcomp, Emesh, flds_scalar_name, flds_scalar_num, rc)
             min_med2mod_areacor_glob, max_med2mod_areacor_glob, 'MOSART'
     end if
 
+    if (fldchk(importState, 'Fgrg_rofl') .and. fldchk(importState, 'Fgrg_rofl')) then
```
See above.

mvertens commented:
@ekluzek @billsacks @slevis-lmwg - I have addressed the issues in this PR except for addressing refactoring the use of  'subname' and reintroducing empty routines whose presence were confusing.

billsacks commented:
It looks like @mvertens has addressed all of the critical points here; if so, it would be great if any non-critical pieces can be deferred to a follow-up PR. @ekluzek and @slevis-lmwg are there any other important needs that you see that need to be addressed before this can be brought in?

ekluzek commented:
Thanks for working on this @mvertens and @billsacks -- yes this covers what we thought was important. This will also make @slevis-lmwg process easier to bring this part in.

ekluzek commented:
One thing I will point out is that the changes as now implemented do require a CMEPS and CISM update. The logical ctl%rof_from_glc won't work correctly to use an older CISM version, because it's set too late. I'll make an issue regarding this and I'll assume that will be fixed in a later PR. It sounds like there is a plan for some follow on PR's as well.

ekluzek commented:
OK I made an issue in #103 about the point I made just above.

slevis-lmwg reviewed (approved):
With one  izumi aux_clm test still in progress, I'm approving this PR. 

The mosart test results appear in the umbrella PR [here ](https://github.com/ESCOMP/CTSM/issues/2590#issuecomment-2179451664)and can be summarized as follows:
- derecho and izumi OK: NLCOMP and "field lists differ"
- new baseline mosart1.1.02_ctsm5.2.007

For the aux_clm results, see the "umbrella" PR.

slevis-lmwg commented:
Made a new tag:
```
git tag -a mosart1.1.02
git push escomp mosart1.1.02
```
